File: BoxBoxNow/BoxBoxNow/ViewModels/ConfigViewModel.swift
import Foundation
import Combine

final class ConfigViewModel: ObservableObject {
    @Published var session: RaceSession = .empty
    @Published var circuits: [Circuit] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    /// True when the user has an active session row on the backend.
    /// When false, saving must go through POST /config/session (create)
    /// instead of PATCH /config/session (update) — otherwise the backend
    /// returns 404 "No active session".
    @Published private(set) var hasActiveSession = false

    func loadSession() async {
        await MainActor.run { isLoading = true }
        do {
            if let s = try await APIClient.shared.getActiveSession() {
                await MainActor.run {
                    self.session = s
                    self.hasActiveSession = true
                    self.isLoading = false
                    print("[Config] Loaded session: kart=\(s.ourKartNumber), circuit=\(s.circuitId ?? -1), duration=\(s.durationMin)")
                }
            } else {
                await MainActor.run {
                    // Keep defaults from .empty so the form is usable.
                    self.hasActiveSession = false
                    self.isLoading = false
                    print("[Config] No active session on server — using defaults; will create on save")
                }
            }
        } catch {
            await MainActor.run {
                self.isLoading = false
                self.errorMessage = error.localizedDescription
            }
            print("[Config] Failed to load session: \(error)")
        }
    }

    func loadCircuits() async {
        do {
            let c = try await APIClient.shared.getMyCircuits()
            await MainActor.run {
                self.circuits = c
                // If we have no session yet and the user hasn't picked a
                // circuit, default to the first accessible one so POST has
                // a valid circuit_id to send.
                if !self.hasActiveSession, self.session.circuitId == nil, let first = c.first {
                    self.session.circuitId = first.id
                    self.session.circuitName = first.name
                }
            }
        } catch {
            print("[Config] Failed to load circuits: \(error)")
        }
    }

    func saveSession() async {
        do {
            let saved: RaceSession
            if hasActiveSession {
                saved = try await APIClient.shared.updateSession(session)
            } else {
                // First-time save — must create the session via POST so the
                // backend has a row to attach subsequent PATCHes to.
                guard session.circuitId != nil else {
                    await MainActor.run {
                        self.errorMessage = "Selecciona un circuito antes de guardar"
                    }
                    return
                }
                saved = try await APIClient.shared.createSession(session)
            }
            await MainActor.run {
                self.session = saved
                self.hasActiveSession = true
                print("[Config] Saved session OK: kart=\(saved.ourKartNumber)")
            }
        } catch {
            await MainActor.run { self.errorMessage = error.localizedDescription }
            print("[Config] Failed to save session: \(error)")
        }
    }
}

File: android/app/src/main/java/com/boxboxnow/app/vm/ConfigViewModel.kt
package com.boxboxnow.app.vm

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.boxboxnow.app.models.Circuit
import com.boxboxnow.app.models.RaceSession
import com.boxboxnow.app.net.ApiClient
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ConfigViewModel @Inject constructor(
    private val api: ApiClient,
) : ViewModel() {
    private val _session = MutableStateFlow(RaceSession.EMPTY)
    val session = _session.asStateFlow()

    private val _circuits = MutableStateFlow<List<Circuit>>(emptyList())
    val circuits = _circuits.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage = _errorMessage.asStateFlow()

    private val _hasActiveSession = MutableStateFlow(false)
    val hasActiveSession = _hasActiveSession.asStateFlow()

    fun loadSession() {
        _isLoading.value = true
        viewModelScope.launch {
            try {
                val s = api.getActiveSession()
                if (s != null) {
                    _session.value = s
                    _hasActiveSession.value = true
                } else {
                    _hasActiveSession.value = false
                }
            } catch (e: Throwable) {
                _errorMessage.value = e.message ?: "Error"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun loadCircuits() {
        viewModelScope.launch {
            runCatching { api.getMyCircuits() }.onSuccess { list ->
                _circuits.value = list
                if (!_hasActiveSession.value && _session.value.circuitId == null) {
                    list.firstOrNull()?.let { first ->
                        _session.value = _session.value.copy(
                            circuitId = first.id,
                            circuitName = first.name,
                        )
                    }
                }
            }
        }
    }

    fun updateSession(transform: (RaceSession) -> RaceSession) {
        _session.value = transform(_session.value)
    }

    fun saveSession() {
        viewModelScope.launch {
            try {
                val saved = if (_hasActiveSession.value) {
                    api.updateSession(_session.value)
                } else {
                    if (_session.value.circuitId == null) {
                        _errorMessage.value = "Selecciona un circuito antes de guardar"
                        return@launch
                    }
                    api.createSession(_session.value)
                }
                _session.value = saved
                _hasActiveSession.value = true
            } catch (e: Throwable) {
                _errorMessage.value = e.message ?: "Error"
            }
        }
    }
}
